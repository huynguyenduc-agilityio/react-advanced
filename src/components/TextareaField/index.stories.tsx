import { Flex } from '@chakra-ui/react';
import { StoryObj, Meta } from '@storybook/react';

import TextareaField from '.';
import { HiMiniPencil } from 'react-icons/hi2';

const defaultProps = {
  label: 'Description',
  name: 'description',
  placeholder: 'Write a short bio about you...',
  icon: HiMiniPencil,
  onChange: () => {},
};

const meta = {
  title: 'Components/TextareaField',
  component: TextareaField,
  parameters: {
    layout: 'centered',
  },

  tags: ['autodocs'],

  render: (args) => (
    <Flex justifyContent="center" w="1000px" p={9} bgColor="spaceBlue">
      <TextareaField {...args} />
    </Flex>
  ),
} satisfies Meta<typeof TextareaField>;

export default meta;

type Story = StoryObj<typeof TextareaField>;

export const Default: Story = {
  args: {
    ...defaultProps,
  },
};

export const WithError: Story = {
  args: {
    ...defaultProps,
    isError: true,
  },
};
