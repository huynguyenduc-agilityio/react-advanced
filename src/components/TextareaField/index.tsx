import { ChangeEvent, memo } from 'react';
import {
  FormLabel,
  FormErrorMessage,
  FormControl,
  Textarea,
  TextareaProps,
  Flex,
  Box,
  Icon,
} from '@chakra-ui/react';
import { IconType } from 'react-icons';

type TTextareaFieldProps = Omit<TextareaProps, 'onChange'> & {
  label: string;
  errorMessages?: string;
  isError?: boolean;
  icon: IconType;
  onChange: (value: string) => void;
};

const TextareaField = ({
  isError = false,
  errorMessages = 'Default error',
  label,
  icon,
  onChange,
  ...rest
}: TTextareaFieldProps) => {
  const handleChangeValue = (e: ChangeEvent<HTMLTextAreaElement>): void =>
    onChange(e.target.value);

  return (
    <FormControl my={6} isInvalid={isError}>
      <Flex align="start">
        <Flex align="center" mt={1}>
          <Icon as={icon} color="white" boxSize={3} mr={1} />
          <FormLabel
            fontSize="sm"
            color="white"
            w="168px"
            textAlign="left"
            mb={0}
          >
            {label}
          </FormLabel>
        </Flex>
        <Box flex="1">
          <Textarea
            variant="primary"
            onChange={handleChangeValue}
            {...rest}
            isInvalid={isError}
          />
          {isError && (
            <FormErrorMessage color="red" fontSize="2xs">
              {errorMessages}
            </FormErrorMessage>
          )}
        </Box>
      </Flex>
    </FormControl>
  );
};

export default memo(TextareaField);
