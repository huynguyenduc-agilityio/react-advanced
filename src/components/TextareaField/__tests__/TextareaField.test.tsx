import { fireEvent, render } from '@testing-library/react';

// Components
import TextareaField from '..';
import { HiMiniPencil } from 'react-icons/hi2';

describe('TextareaField component', () => {
  const mockOnChange = jest.fn();
  const props = {
    label: 'Description',
    name: 'description',
    icon: HiMiniPencil,
    placeholder: 'Description',
    onChange: mockOnChange,
  };
  it('should match snapshot', () => {
    const { container } = render(<TextareaField {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should call onChange when input value is changed', async () => {
    const { getByPlaceholderText } = render(<TextareaField {...props} />);
    fireEvent.change(getByPlaceholderText('Description'), {
      target: { value: 'user@gmail.com' },
    });

    expect(mockOnChange).toHaveBeenCalled();
  });

  it('should display error message when isError is true', () => {
    const { getByText } = render(
      <TextareaField {...props} isError errorMessages="Required field" />,
    );

    expect(getByText('Required field')).toBeInTheDocument();
  });
});
