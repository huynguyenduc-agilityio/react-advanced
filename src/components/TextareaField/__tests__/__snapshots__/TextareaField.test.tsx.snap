// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaField component render correctly with default props 1`] = `
<div>
  <div
    class="chakra-form-control css-zol2i8"
    role="group"
  >
    <div
      class="css-h38yef"
    >
      <div
        class="css-jes1ca"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-y0hwlo"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m2.695 14.762-1.262 3.155a.5.5 0 0 0 .65.65l3.155-1.262a4 4 0 0 0 1.343-.886L17.5 5.501a2.121 2.121 0 0 0-3-3L3.58 13.419a4 4 0 0 0-.885 1.343Z"
          />
        </svg>
        <label
          class="chakra-form__label css-1jv67cy"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          Description
        </label>
      </div>
      <div
        class="css-1rr4qq7"
      >
        <textarea
          class="chakra-textarea css-0"
          id="field-:r0:"
          name="description"
          placeholder="Description"
        />
      </div>
    </div>
  </div>
</div>
`;
